### DISCUSSION TITLE
Am I missing a sveltekit config?

### DISCUSSION THREAD

**Discussion by longrunningprocess** [saadeghi/daisyui#656] [Q&A] [answered]
I'm getting a bunch of these `Unused CSS selector` warnings when I run my sveltekit builds, e.g., `Unused CSS selector "button-group > .btn:not(:first-of-type)"`.

Do you all see those in your projects or did you add some additional configuration to have them purged during the build?

Thanks!

I've narrowed it down to the way I use `@apply` within my `style` blocks.

When I do things like this I do not see any errors:

```html
<label for={id} class=label>
	<span class='label-text'>{label}</span>
</label>

<input {id} use:typeWorkaround bind:value {required} bind:this={input} class='input input-bordered' />
```

However, when I do it like this, I see a bunch of `Unused CSS selector` messages from `vite-plugin-svelte`:

```html
<label for={id}>
	<span>{label}</span>
</label>

<input {id} use:typeWorkaround bind:value {required} bind:this={input} />

<style>
	/* https://daisyui.com/components/input */
	label { @apply
		label;
	}
	label > span { @apply
		label-text;
	}
	input { @apply
		input
		input-bordered;
	}
</style>

```

**Comment by saadeghi** [comment, marked as answer]
I tried to reproduce this issue but everything seems to work fine:

https://stackblitz.com/edit/daisyui-sveltekit-69ehca?file=src/routes/index.svelte

**Reply by longrunningprocess** [reply]
Ok, I'll take a look later today and compare versions, configs, etc... must be something unique on my side.

**Reply by longrunningprocess** [reply]
ok, I did reproduce this... I wonder if it something to do with the `preprocess` config...

https://stackblitz.com/edit/daisyui-sveltekit-kyzxwt

**Reply by longrunningprocess** [reply]
yep, I removed the `preprocess` lines from the `svelte.config.js` and I don't see those errors anymore.  I originally added that `preprocess` to get the `@apply` to work at all.

**Reply by longrunningprocess** [reply]
since it doesn't appear I need the `preprocess` anymore it has me wondering whether some of my other deps are still needed like `postcss` and `postcss-load-config`, do you put those deps into your Svelte projects?

**Reply by saadeghi** [reply]
I'm not sure 🤔 the preprocess package says it adds support for PostCSS, etc. But if that's required, how is my repo working without it 😅

I usually have 1 CSS file (Tailwind). If I have to scope CSS on a component, it's usually a few normal CSS lines. So I never needed PostCSS on scoped style

**Reply by longrunningprocess** [reply]
🤔 yeah, I'm not sure either... I just removed all of those and everything seems to function for `npm run dev` as well as `npm run build` so I guess I'll just pull them out.  I'm a little befuddled though because I definitely could not use `@apply` with my `style` bocks in a `.svelte` file before I put those extra config's in place so I have to assume something must've been "fixed" in later releases that affects that stuff.

Thanks again, the takeaway here I suppose is that less config is needed to get things up and running 🎉

**Reply by longrunningprocess** [reply]
@saadeghi , I completely forgot this weekend why I had originally put that `preprocess` in place... modifiers won't work in those `@apply` blocks without it.

I've updated https://stackblitz.com/edit/daisyui-sveltekit-diijhg?file=src/routes/index.svelte so you can see the failure that occurs without `preprocess`.

**Reply by longrunningprocess** [reply]
and here's a repro:  https://stackblitz.com/edit/daisyui-sveltekit-kwyhz6?file=src/routes/index.svelte with all of the "Unused" errors I mentioned initially.

**Reply by longrunningprocess** [reply]
I can create a new discussion if you like since I muddied the water on this one by forgetting why I need the `preprocess`.

**Reply by longrunningprocess** [reply]
I mentioned this required sveltekit setup to the Tailwind team:  https://github.com/tailwindlabs/tailwindcss.com/issues/1221

**Reply by saadeghi** [reply]
I see...  
Let's see what's the issue here:  
daisyUI has this selector: `.label a:hover`  
and if we don't have a `a` inside `.label` it's up to the CSS processor that remove this line or keep it.  
Packages like `purgecss` or `uncss` will remove this line if `.label` has no `a` in it but Tailwind CSS's JIT doesn't remove it.  
I think it's because of compatibility reasons. If we have nested components, `.label` is one file and `a` is in another file, we wouldn't want that style to be removed just because the CSS preprocessor didn't find the exact hierarchy of styles in a file.

So, the CSS preprocessor is keeping `.label a:hover` style. then you use `@apply label` as a scoped style inside your Svelte component. what happens here is, Tailwind gives `.label a:hover` style to your component.  
Your component checks the style and compares with its own markup. it finds `label` but can't find `<a>` inside it, so it warns you that you have unused scoped styles.

That's it. and I don't think there's a fix for this.

However, may I ask why are you scoping CSS?  
Because the whole point of using a utility first CSS is avoiding scoped CSS so our class names will be reusable on all components.

**Reply by longrunningprocess** [reply]
wow, that's excellent research, thank you!

> However, may I ask why are you scoping CSS?
Because the whole point of using a utility first CSS is avoiding scoped CSS so our class names will be reusable on all components.

Simply to keep the HTML as clean as possible honestly.  You're already doing so much of that with your work in daisy but I had hoped I'd be able to keep everything in the style blocks.  It doesn't have to be a big deal, I can simply run the tradeoffs by the team and see what they prefer.  I'm going to recommend we put the classes in the HTML and just make it the very last attribute.
